core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java

```diff
@@ -252,76 +258,38 @@ private boolean gainPrimacy() throws Exception {
         if (Configuration.getBoolean(PropertyKey.MASTER_JOURNAL_EXIT_ON_DEMOTION)) {
           stop();
         } else {
-          losePrimacy();
+          demote();
         }
         return false;
       }
     }
     try {
-      startMasters(true);
+      startMasterComponents(true);
     } catch (UnavailableException e) {
       LOG.warn("Error starting masters: {}", e.toString());
       mJournalSystem.losePrimacy();
       stopMasters();
       return false;
     }
-    mServingThread = new Thread(() -> {
-      try {
-        startCommonServices();
-        startMasterServing(" (gained leadership)", " (lost leadership)");
-      } catch (Throwable t) {
-        Throwable root = Throwables.getRootCause(t);
-        if (root instanceof InterruptedException || Thread.interrupted()) {
-          return;
-        }
-        ProcessUtils.fatalError(LOG, t, "Exception thrown in main serving thread");
-      }
-    }, "MasterServingThread");
-    LOG.info("Starting a server thread.");
-    mServingThread.start();
-    if (!waitForReady(10 * Constants.MINUTE_MS)) {
-      ThreadUtils.logAllThreads();
-      throw new RuntimeException("Alluxio master failed to come up");
-    }
+    mServices.forEach(SimpleService::promote);
     LOG.info("Primary started");
     return true;
   }
 
-  private void losePrimacy() throws Exception {
+  private void demote() throws Exception {
     LOG.info("Losing the leadership.");
-    if (mServingThread != null) {
-      stopServingGrpc();
-      stopCommonServicesIfNecessary();
-    }
+    mServices.forEach(SimpleService::demote);
     // Put the journal in standby mode ASAP to avoid interfering with the new primary. This must
     // happen after stopServing because downgrading the journal system will reset master state,
     // which could cause NPEs for outstanding RPC threads. We need to first close all client
     // sockets in stopServing so that clients don't see NPEs.
     mJournalSystem.losePrimacy();
-    if (mServingThread != null) {
```

Comment:
Actually here it gets a bit more interesting if the `mServingThread` is gone. Does that mean the starting/stopping will be done in the main thread and will be essentially blocking for `main` thread? My gut feeling says it's alright because all the RPC/web/metrics requests are handled by dedicated threads and does not need the main thread. But if possible can you try sending some RPC/HTTP requests to the master during this promote/demote and see if there's anything weird?

Reply:
I bombarded the master during promote/demote with HTTP calls to gather its metrics as well as RPC calls to create files and saw nothing out of the ordinary.